### PR TITLE
Adding pointcloud

### PR DESCRIPTION
**Pull request by monfera**
This PR contains the plotly.js part of the new `pointcloud2d` trace type.

Work left involves

- [x] changing certain things about sizing
- [x] removing some disused marker options
- [x] image test cases
- [x] npm publish `gl-pointcloud2d`

### CODE REVIEW

File: lib/index-gl2d.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ var Plotly = require('./core');
 
 Plotly.register([
     require('./scattergl'),
+    require('./pointcloud2d'),
```
Comment (etpinard) [inline lib/index-gl2d.js]: Let's make this `pointcloud` !

File: src/traces/pointcloud2d/convert.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        name: this.name,
+        hoverinfo: this.hoverinfo
+    };
+};
+
+/**
+ * Truncate a Float32Array to some length. A wrapper to support environments
+ * (e.g. node-webkit) that do not implement Float32Array.prototype.slice
+ */
+function truncate(float32ArrayIn, len) {
```
Comment (etpinard) [inline src/traces/pointcloud2d/convert.js]: :palm_tree: we should make this a `src/lib/` function.

File: src/traces/pointcloud2d/convert.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    this.pointcloud.dispose();
+};
+
+function createPointcloud(scene, data) {
+    var plot = new Pointcloud(scene, data.uid);
+    plot.update(data);
+    return plot;
+}
+
+module.exports = createPointcloud;
```
Comment (etpinard) [inline src/traces/pointcloud2d/convert.js]: Nicely done.

File: lib/index-gl2d.js (hunk 1 above)
Comment (monfera) [inline lib/index-gl2d.js]: OK, now done

File: src/traces/pointcloud/attributes.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+*/
+
+'use strict';
+
+var scatterglAttrs = require('../scattergl/attributes');
+
+module.exports = {
+    x: scatterglAttrs.x,
+    y: scatterglAttrs.y,
+    xy: {
```
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: _a dream come true_

File: src/traces/pointcloud/attributes.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        valType: 'data_array',
+        description: [
+            'Faster alternative to specifying `x` and `y` separately.',
+            'If supplied, it must be a typed `Float32Array` array that',
+            'represents points such that `xy[i * 2] = x[i]` and `xy[i * 2 + 1] = y[i]`'
+        ].join(' ')
+    },
+    indexid: {
```
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: @monfera do the `indexid` items have to be numbers? 

If so, I'd vote for naming this attribute `indices`. If not, I'd

File: src/traces/pointcloud/attributes.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        valType: 'data_array',
+        description: [
+            'A sequential value, 0..n, supply it to avoid creating this array inside plotting.',
+            'Its length must be equal to or greater than the number of points.',
+            'For the best performance and memory use, create one large `indexid` typed array',
+            'that is guaranteed to be at least as long as the largest number of points during',
+            'use, and reuse it on each `Plotly.restyle()` call.'
+        ].join(' ')
+    },
+    bounds: {
```
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: perhaps, splitting this into `xbounds` and `ybounds` would a little more plotly-esque.

cc @cldougl

File: src/traces/pointcloud/attributes.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            role: 'style',
+            description: [
+                'Sets the marker opacity. The default value is `1` (fully opaque).',
+                'If the markers are not fully opaque and there are hundreds of thousands',
+                'of points, it may cause slower zooming and panning.',
+                'Opacity fades the color even if `blend` is left on `false` even if there',
+                'is no translucency effect in that case.'
+            ].join(' ')
+        },
+        blend: {
```
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: nice :tada:

File: src/traces/pointcloud/attributes.js (hunk 5 above)
Comment (monfera) [inline src/traces/pointcloud/attributes.js]: Yes, they have to be numbers.

Comment (rreusser) [inline src/traces/pointcloud/attributes.js]: As discussed, animation ids must be strings.

File: src/traces/pointcloud/attributes.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            valType: 'number',
+            min: 0.1,
+            dflt: 20,
+            role: 'style',
+            description: [
+                'Sets the maximum size (in px) of the rendered marker points.',
+                'Effective when the `pointcloud` shows only few points.'
+            ].join(' ')
+        },
+        border: {
```
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: Yes, `marker.border` is different enough from `marker.line` to deserve its own name :+1:

File: src/traces/pointcloud/attributes.js (hunk 5 above)
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: Ok. Let's go with `indices` then :+1:

File: src/traces/pointcloud/convert.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var createpointcloud = require('gl-pointcloud2d');
```
Comment (etpinard) [inline src/traces/pointcloud/convert.js]: :camel: case please

File: src/traces/pointcloud/convert.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        positions = xy;
+
+        if(userBounds) {
+
+            bounds[0] = userBounds[0];
+            bounds[1] = userBounds[1];
+            bounds[2] = userBounds[2];
+            bounds[3] = userBounds[3];
+
+        } else {
```
Comment (etpinard) [inline src/traces/pointcloud/convert.js]: Great. Nice job opening this fast path :racehorse:

File: src/traces/pointcloud/attributes.js (hunk 6 above)
Comment (monfera) [inline src/traces/pointcloud/attributes.js]: Just checked, there's no obstacle to it, I can switch it over.

File: src/traces/pointcloud/defaults.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+module.exports = function supplyDefaults(traceIn, traceOut, defaultColor) {
+    function coerce(attr, dflt) {
+        return Lib.coerce(traceIn, traceOut, attributes, attr, dflt);
+    }
+
+    coerce('x');
+    coerce('y');
+    coerce('bounds');
+
+    if(traceIn.xy && traceIn.xy instanceof Float32Array) {
```
Comment (etpinard) [inline src/traces/pointcloud/defaults.js]: We'll abstract this logic out at some point on our way to https://github.com/plotly/plotly.js/issues/860

But, this is great for now :+1:

File: src/lib/float32_truncate.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+'use strict';
+
+/**
+ * Truncate a Float32Array to some length. A wrapper to support environments
+ * (e.g. node-webkit) that do not implement Float32Array.prototype.slice
+ */
+module.exports = function truncate(float32ArrayIn, len) {
+    // for some reason, ES2015 Float32Array.prototype.slice takes 2x as long...
+    // therefore we aren't checking for its existence
```
Comment (etpinard) [inline src/lib/float32_truncate.js]: Very interesting comment. Thanks for checking!

Comment (monfera) [inline src/lib/float32_truncate.js]: All too often, 'recently' added language elements (as in: the last few years) are slower even if intuitively they should be faster. Back in 2014, untyped arrays were faster than typed arrays even for plain calculations; same with `Map`.

**Comment by etpinard** [comment]
@monfera Looks like the the autoscale and reset scale mode bar buttons don't work 

gifrecord_2016-09-01_180432

I suspect that caused by the `doAutorange` not finding all the required info. 

We might want to have a look at https://github.com/plotly/plotly.js/pull/855/commits/330c5f3cd1aaccad9fa1c9607006f5891dd78c69

**Comment by etpinard** [comment]
Apart from the above comment and a few naming / styling issues this PR is looking amazing. 

@monfera Thanks for all your hard work!

File: src/traces/pointcloud/attributes.js (hunk 6 above)
Comment (cldougl) [inline src/traces/pointcloud/attributes.js]: 👍 👍 definitely prefer `xbounds` and `ybounds`

**Comment by monfera** [comment]
@etpinard thanks for all your comments, they look easy, i.e. perfect for a morning ramp-up. Special thanks for catching this interaction issue too!

**Comment by monfera** [comment]
@etpinard I just wanted to mention that I had some unpushed commits in `gl-gis/pointcloud-2d` and the image now should come out like this (see the difference in the opacity, size and boundary). 
image

**Comment by etpinard** [comment]
dropping this PR from the `v1.17.0` pipeline because of the :no_entry_sign: https://github.com/stackgl/gl-shader/pull/16

**Comment by monfera** [comment]
@etpinard for the record the solution to the plot disappearance issue on +/-/reset you found is here in `stackgl`: https://github.com/stackgl/gl-shader/pull/16 - thanks again for catching it!

File: src/traces/pointcloud/convert.js [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        idToIndex,
+        positions,
+        bounds = this.bounds;
+
+    var xx, yy, i;
+
+    if(xy) {
+
+        positions = xy;
+        len = xy.length >>> 1;
```
Comment (etpinard) [inline src/traces/pointcloud/convert.js]: bitwise ops are cool :sunglasses: but can you add a short comment above describing what this does (for folks unfamiliar with them).

File: src/traces/pointcloud/convert.js [hunk 14]
```diff
@@ -96,6 +96,8 @@ proto.updateFast = function(options) {
     if(xy) {
 
         positions = xy;
+
+        // dividing xy.length by 2 and truncating to integer if xy.length was not even
```
Comment (etpinard) [inline src/traces/pointcloud/convert.js]: thank you very much!

**Comment by etpinard** [comment]
Amazing work :dancer:

**Pull request merged**